File: apps/web/lib/api/llm.ts
import { request } from "@/lib/http";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse） */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

/** 供应商类型（与后端 movieclaw_llm 预设 id 对应）。 */
export type LlmProviderType =
  | "openai"
  | "bailian"
  | "deepseek"
  | "kimi"
  | "glm"
  | "openai_compat";

/** 连接验证状态（与站点/下载器共用同一状态机语义）。 */
export type LlmProviderStatus = "pending" | "verifying" | "active" | "failed";

/** 预设模型目录条目（见 movieclaw_llm.models.ModelInfo）。 */
export interface LlmModelInfo {
  id: string;
  /** 输入+输出共享的总上下文（token）；null 表示官方未公布 */
  context_window: number | null;
  /** 单独的输入上限（百炼公布，OpenAI 不单独公布） */
  max_input_tokens: number | null;
  /** 单次响应的输出上限 */
  max_output_tokens: number | null;
  supports_tools: boolean;
  /** 是否支持一次响应发起多个工具调用 */
  supports_parallel_tool_calls: boolean;
  /** 是否会输出思考内容（reasoning_content） */
  supports_thinking: boolean;
  /** 思维链预算上限（thinking_budget 最大值）；null = 不支持或未公布 */
  max_thinking_tokens: number | null;
  modalities: string[];
}

/** 供应商预设（见 schemas.llm.LlmPresetView）：设置页渲染选项用。 */
export interface LlmPreset {
  id: LlmProviderType;
  display_name: string;
  /** 预设默认端点；null 表示走官方默认或必须自填 */
  base_url: string | null;
  /** 是否必须填写 base_url（通用兼容端点没有默认值） */
  requires_base_url: boolean;
  models: LlmModelInfo[];
}

/** 当前配置的对外视图（见 schemas.llm.LlmProviderView，脱敏无 API Key）。 */
export interface LlmProviderConfig {
  provider_type: LlmProviderType;
  base_url: string | null;
  default_model: string;
  status: LlmProviderStatus;
  /** 是否可用 = 连接测试通过 */
  usable: boolean;
  last_error: string | null;
  last_checked_at: string | null;
  /** 最近验证成功时端点上报的可用模型列表 */
  available_models: string[] | null;
  /** 用户补录的自定义模型目录（含参数），设置页下拉框的数据源之一 */
  extra_models: LlmModelInfo[];
  created_at: string;
  updated_at: string;
}

/** 保存配置的请求体（见 schemas.llm.LlmProviderPayload）。 */
export interface LlmProviderPayload {
  provider_type: LlmProviderType;
  base_url?: string | null;
  api_key: string;
  default_model: string;
  /** 自定义模型目录：default_model 不在预设目录时，必须在这里带齐参数 */
  extra_models?: LlmModelInfo[];
}

/** 列出可接入的供应商类型及其模型目录。 */
export function listLlmPresets(init?: RequestInit): Promise<LlmPreset[]> {
  return unwrap(request<ApiEnvelope<LlmPreset[]>>("/llm/presets", init));
}

/** 获取当前配置；尚未配置时返回 null（用于轮询连接测试进度）。 */
export function getLlmProvider(init?: RequestInit): Promise<LlmProviderConfig | null> {
  return unwrap(request<ApiEnvelope<LlmProviderConfig | null>>("/llm/provider", init));
}

/** 保存配置（单例 upsert，保存后后端异步测试连接）。 */
export function saveLlmProvider(payload: LlmProviderPayload): Promise<LlmProviderConfig> {
  return unwrap(
    request<ApiEnvelope<LlmProviderConfig>>("/llm/provider", {
      method: "PUT",
      body: JSON.stringify(payload),
    }),
  );
}

/** 手动重新测试一次连接。 */
export function reverifyLlmProvider(): Promise<LlmProviderConfig> {
  return unwrap(
    request<ApiEnvelope<LlmProviderConfig>>("/llm/provider/verify", { method: "POST" }),
  );
}

/** 删除配置。 */
export function deleteLlmProvider(): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>("/llm/provider", { method: "DELETE" }),
  );
}

File: src/movieclaw_api/api/routes/llm.py
from __future__ import annotations

from fastapi import APIRouter, BackgroundTasks, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from movieclaw_api.schemas.llm import LlmPresetView, LlmProviderPayload, LlmProviderView
from movieclaw_api.schemas.response import ApiResponse, ok
from movieclaw_api.services.llm_config import LlmConfigService, verify_llm_provider
from movieclaw_db.engine import get_session
from movieclaw_llm.providers import list_presets

router = APIRouter(prefix="/llm", tags=["llm"])


@router.get(
    "/presets",
    response_model=ApiResponse[list[LlmPresetView]],
    summary="列出可接入的供应商类型及其模型目录",
)
async def list_llm_presets() -> ApiResponse[list[LlmPresetView]]:
    """返回内置供应商预设（OpenAI / 阿里云百炼 / 通用兼容端点），
    设置页据此渲染类型选项、端点默认值与模型选择提示。"""
    return ok([LlmPresetView.from_preset(p) for p in list_presets()])


@router.get(
    "/provider",
    response_model=ApiResponse[LlmProviderView | None],
    summary="获取当前的模型供应商配置",
)
async def get_llm_provider(
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[LlmProviderView | None]:
    """单例配置：尚未配置时 data 为 null（不是 404），设置页据此渲染空态。"""
    service = LlmConfigService(session)
    row = await service.get_or_none()
    return ok(LlmProviderView.from_model(row) if row else None)


@router.put(
    "/provider",
    response_model=ApiResponse[LlmProviderView],
    summary="保存模型供应商配置（保存后异步测试连接）",
)
async def save_llm_provider(
    payload: LlmProviderPayload,
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[LlmProviderView]:
    """单例 upsert：有配置则整体覆盖。保存后状态置 pending 并在后台
    用默认模型发一次最小对话验证；前端可轮询 GET /llm/provider 观察
    status：pending → verifying → active / failed（见 last_error）。"""
    service = LlmConfigService(session)
    await service.upsert(
        provider_type=payload.provider_type,
        base_url=payload.base_url,
        api_key=payload.api_key,
        default_model=payload.default_model,
        extra_models=payload.extra_models,
    )
    row = await service.start_verification()
    background_tasks.add_task(verify_llm_provider)
    return ok(LlmProviderView.from_model(row), message="已保存，正在测试模型连接")


@router.post(
    "/provider/verify",
    response_model=ApiResponse[LlmProviderView],
    summary="手动重新测试模型连接",
)
async def reverify_llm_provider(
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[LlmProviderView]:
    """行为：未配置 → 404；已在测试中 → 409；否则同步占位为 VERIFYING
    并在后台重新测试。"""
    service = LlmConfigService(session)
    row = await service.start_verification()
    background_tasks.add_task(verify_llm_provider)
    return ok(LlmProviderView.from_model(row), message="已重新发起连接测试")


@router.delete(
    "/provider",
    response_model=ApiResponse[dict],
    summary="删除模型供应商配置",
)
async def delete_llm_provider(
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[dict]:
    service = LlmConfigService(session)
    await service.delete()
    return ok({}, message="已删除")
